pkg/pillar/cmd/zedmanager/handlevolumemgr.go

```diff
@@ -6,173 +6,171 @@ package zedmanager
+// MaybeAddVolumeRefConfig publishes volume ref config with refcount
+// to the volumemgr
+func MaybeAddVolumeRefConfig(ctx *zedmanagerContext, appInstID uuid.UUID,
+	volumeID uuid.UUID, generationCounter int64) {
 
-	key := types.VolumeKeyFromParts(blobSha256, appInstID, volumeID, purgeCounter)
-	log.Infof("AddOrRefcountVolumeConfig for %s", key)
-	m := lookupVolumeConfig(ctx, key)
+	key := fmt.Sprintf("%v#%d", volumeID, generationCounter)
+	log.Infof("MaybeAddVolumeRefConfig for %s", key)
+	m := lookupVolumeRefConfig(ctx, key)
 	if m != nil {
 		m.RefCount++
-		log.Infof("VolumeConfig exists for %s to refcount %d",
+		log.Infof("VolumeRefConfig exists for %s to refcount %d",
 			key, m.RefCount)
-		publishVolumeConfig(ctx, m)
+		publishVolumeRefConfig(ctx, m)
 	} else {
-		log.Debugf("AddOrRefcountVolumeConfig: add for %s", key)
-		// XXX hard-coded for OriginTypeDownload for now
-		d := types.DownloadOriginConfig{
-			ImageID:     ss.ImageID,
-			DatastoreID: ss.DatastoreID,
-			Name:        ss.Name,
-			NameIsURL:   ss.NameIsURL,
-			ImageSha256: ss.ImageSha256,
-			IsContainer: ss.IsContainer,
-			AllowNonFreePort: types.AllowNonFreePort(*ctx.globalConfig,
-				types.AppImgObj),
-			MaxDownSize: ss.MaxDownSize,
-			FinalObjDir: ss.FinalObjDir, // XXX use?
-
-			CertificateChain: ss.CertificateChain,
-			ImageSignature:   ss.ImageSignature,
-			SignatureKey:     ss.SignatureKey,
-		}
-		n := types.OldVolumeConfig{
-			BlobSha256:     blobSha256,
-			AppInstID:      appInstID,
-			VolumeID:       volumeID,
-			PurgeCounter:   purgeCounter,
-			DisplayName:    ss.Name,
-			Origin:         types.OriginTypeDownload,
-			DownloadOrigin: &d,
-			MaxVolSize:     ss.MaxVolSize,
-			ReadOnly:       ss.ReadOnly,
-			Format:         ss.Format,
-			Target:         ss.Target,
-			RefCount:       1,
+		log.Debugf("MaybeAddVolumeRefConfig: add for %s", key)
+		vrc := types.VolumeRefConfig{
+			VolumeID:          volumeID,
+			GenerationCounter: generationCounter,
+			RefCount:          1,
 		}
-		publishVolumeConfig(ctx, &n)
+		publishVolumeRefConfig(ctx, &vrc)
 	}
 	base.NewRelationObject(base.AddRelationType, base.AppInstanceConfigLogType, appInstID.String(),
-		base.OldVolumeConfigLogType, key).Infof("App instance to volume relation.")
-	log.Infof("AddOrRefcountVolumeConfig done for %s", key)
+		base.VolumeRefConfigLogType, key).Infof("App instance to volume relation.")
+	log.Infof("MaybeAddVolumeRefConfig done for %s", key)
 }
 
-// MaybeRemoveVolumeConfig decreases the RefCount and deletes the VolumeConfig
+// MaybeRemoveVolumeRefConfig decreases the RefCount and deletes the VolumeRefConfig
 // when the RefCount reaches zero
-func MaybeRemoveVolumeConfig(ctx *zedmanagerContext, blobSha256 string,
-	appInstID uuid.UUID, volumeID uuid.UUID, purgeCounter uint32) {
+func MaybeRemoveVolumeRefConfig(ctx *zedmanagerContext, appInstID uuid.UUID,
+	vrs types.VolumeRefStatus) {
```

Comment:
It is easier to maintain (less risk of this function doing stuff to the vrs) if you make it take the same arguments as the Add function i.e.,
	volumeID uuid.UUID, generationCounter int64) {
then
		key := fmt.Sprintf("%v#%d", volumeID, generationCounter)

Reply:
Resolved.